### ISSUE TITLE
Check for conflicting usernames in register

### ISSUE THREAD

**Issue by smart-reports** [staart/api#1057]
Hi Anand, here is the behaviour we're seeing now

User registers and verifies email - works fine now
User tries to login - immediately presented with a "You dont have permission to perform this action"....
Press F5 to refresh and it looks like the user is logged in - 
However, when I start to click on the user icon the top right and select team or user settings - it keeps presenting us with this "You dont have permission".  

Must be something we are not doing?

![image](https://user-images.githubusercontent.com/50381700/77312018-1862b500-6cf9-11ea-836d-d46c07837775.png)

**Comment by AnandChowdhary** [comment]
The really interesting part is that I can still log in and it works. It has something to do with creating new users. Perhaps they don't get the right permissions etc.

**Comment by AnandChowdhary** [comment]
I've added some logging for the authorization helper so I can see what's going wrong on my server: https://github.com/staart/api/commit/347ec517759109545f1e3d98091b3e7197bc45f5

**Comment by AnandChowdhary** [comment]
For my account (which was created a long time ago and works perfectly), the logs are :

```
2020-03-23T12:37:00.537440723Z Checking authorization 1 user:memberships:read 1
2020-03-23T12:37:00.540051314Z Checking user authorization RowDataPacket {
2020-03-23T12:37:00.540157141Z updatedAt: '2020-03-20T16:55:20.000Z' } user:memberships:read RowDataPacket {
2020-03-23T12:37:00.540240233Z updatedAt: '2020-03-20T16:55:20.000Z' }
2020-03-23T12:37:00.541335231Z Is user target? true
2020-03-23T12:37:00.541346856Z Is user admin? false
```

For the broken account (ogremagi07@gmail.com), the logs are:

```
2020-03-23T12:39:26.403918478Z Checking authorization 32 user:memberships:read 22
2020-03-23T12:39:26.404282230Z Checking user authorization RowDataPacket {
2020-03-23T12:39:26.404297435Z id: '32',
2020-03-23T12:39:26.404302300Z name: 'Gino B',
2020-03-23T12:39:26.404306387Z username: 'gino',
2020-03-23T12:39:26.404310272Z nickname: 'Gino',
2020-03-23T12:39:26.404314166Z primaryEmail: '32',
2020-03-23T12:39:26.404318164Z twoFactorEnabled: false,
2020-03-23T12:39:26.404332343Z countryCode: 'us',
2020-03-23T12:39:26.404336067Z timezone: 'Europe/Amsterdam',
2020-03-23T12:39:26.404339461Z notificationEmails: '2',
2020-03-23T12:39:26.404342975Z preferredLanguage: 'en-us',
2020-03-23T12:39:26.404346340Z prefersReducedMotion: false,
2020-03-23T12:39:26.404349754Z prefersColorSchemeDark: false,
2020-03-23T12:39:26.404353217Z checkLocationOnLogin: false,
2020-03-23T12:39:26.404356753Z role: '1',
2020-03-23T12:39:26.404360070Z gender: 'x',
2020-03-23T12:39:26.404363442Z profilePicture:
2020-03-23T12:39:26.404366981Z 'https://www.gravatar.com/avatar/4b4aed1f68d3fd9d564f82abbc44852e?d=https%3A%2F%2Fapi.adorable.io%2Favatars%2F285%2F148f99332d830fda1a795d7728ce882c.png',
2020-03-23T12:39:26.404371252Z createdAt: '2020-03-23T11:39:53.000Z',
2020-03-23T12:39:26.404374700Z updatedAt: '2020-03-23T11:39:53.000Z' } user:memberships:read RowDataPacket {
2020-03-23T12:39:26.404378244Z id: '22',
2020-03-23T12:39:26.404381580Z name: 'Gino Gino',
2020-03-23T12:39:26.404385119Z username: 'gino',
2020-03-23T12:39:26.404398279Z nickname: 'Gino',
2020-03-23T12:39:26.404411098Z primaryEmail: '22',
2020-03-23T12:39:26.404414496Z twoFactorEnabled: false,
2020-03-23T12:39:26.404418063Z countryCode: 'us',
2020-03-23T12:39:26.404421391Z timezone: 'Europe/Amsterdam',
2020-03-23T12:39:26.404424855Z notificationEmails: '2',
2020-03-23T12:39:26.404428291Z preferredLanguage: 'en-us',
2020-03-23T12:39:26.404431909Z prefersReducedMotion: false,
2020-03-23T12:39:26.404435410Z prefersColorSchemeDark: false,
2020-03-23T12:39:26.404438890Z checkLocationOnLogin: false,
2020-03-23T12:39:26.404442263Z role: '1',
2020-03-23T12:39:26.404445717Z gender: 'x',
2020-03-23T12:39:26.404449050Z profilePicture:
2020-03-23T12:39:26.404452511Z 'https://www.gravatar.com/avatar/fc7364ae5ac620c61ac65fc2934539b9?d=https%3A%2F%2Fapi.adorable.io%2Favatars%2F285%2Fe1c182f5baf4a1630aeb0aa979e9686d.png',
2020-03-23T12:39:26.404456239Z createdAt: '2020-03-22T12:13:17.000Z',
2020-03-23T12:39:26.404459839Z updatedAt: '2020-03-22T12:13:17.000Z' }
2020-03-23T12:39:26.404639895Z Is user target? false
2020-03-23T12:39:26.404649289Z Is user admin? false
2020-03-23T12:39:26.404652775Z Got user memberships []
2020-03-23T12:39:26.404656096Z Got target memberships []
2020-03-23T12:39:26.404659441Z Got similar memberships []
2020-03-23T12:39:26.404662888Z Allowed? false
```

**Comment by AnandChowdhary** [comment]
So I see that there are two users with the username "gino", one with user ID 32 and another with 22. When you log in, we make a GET request to something like `/users/gino` instead of using your user ID, and that conflict means you're trying to request the information of another user, which you don't have the permission to.

The interesting part here is how you were able to make two accounts with the same username. I'm fairly confident it has something to do with my most recent update that creates "best looking" usernames, called `getBestUsernameForUser`. However, if you supply the `username` key when creating an account, I also see that we don't check for collisions, which is very bad.

These are the problem causing lines: https://sourcegraph.com/github.com/staart/api/-/blob/src/rest/auth.ts#L128-130

**Comment by AnandChowdhary** [comment]
There are 4 `gino`s and 2 `testing`s and `stuart`s, which highlights that the logic for checking conflicting usernames is broken. I'll get a fix running ASAP, and for now I'm manually changing the usernames of the two conflicting users. This can also be fixed by marking `username` as a unique key in the database.

**Comment by smart-reports** [comment]
how do i set myself to an admin user?

**Comment by smart-reports** [comment]
thanks for your help Anand and resolving these issues today

**Comment by AnandChowdhary** [comment]
> how do i set myself to an admin user?

You'll have to change the user directly in the database. Change `role` to `3` (admin).

**Comment by AnandChowdhary** [comment]
This is fixed in `v1.3.157`.  I feel so stupid! The mistake was missing `await` when checking availability, evaluating `Promise<pending>` == `{}` to be `true`.
